File: mover/lizhecome/code/task8/get_proof.py
import hashlib
import random

def sha3_256(data: bytes) -> bytes:
    """计算 sha3_256 哈希"""
    return hashlib.sha3_256(data).digest()

def find_proof(sender_bytes: bytes, challenge_bytes: bytes, difficulity: int = 3) -> list:
    """寻找满足条件的 proof 并返回字节数组"""
    while True:
        # 随机生成 8 字节的 proof
        proof = random.randbytes(8)
        
        # 构建 full_proof = proof + sender_bytes + challenge_bytes
        full_proof = proof + sender_bytes + bytes(challenge_bytes)
        
        # 计算 sha3_256 哈希
        hash_result = sha3_256(full_proof)
        
        # 计算前 difficulity 个字节的前缀和
        prefix_sum = sum(hash_result[i] for i in range(difficulity))
        
        # 满足 prefix_sum == 0 的条件
        if prefix_sum == 0:
            print(f"Proof found (byte array): {[b for b in proof]}")
            return [b for b in proof]

# 将 sender 转换为字节数组（16 进制字符串转为 bytes）
sender_hex = "43d945f82670c017d1989a6613612092e07560dcf88580f6649c4c0b7aa54e44"
sender_bytes = bytes.fromhex(sender_hex)

# 使用提供的 challenge_bytes
challenge_bytes = [
    25, 231, 108, 165, 4, 197, 165, 250, 94, 33, 74, 69, 252, 166, 192, 88,
    23, 27, 163, 51, 246, 218, 137, 123, 130, 115, 16, 148, 80, 77, 90, 185,
    24, 53, 123, 75, 62, 115, 100, 82, 75, 76, 48, 74, 44, 60, 65, 40, 94, 45,
    126, 115, 77, 126, 95, 121, 51, 3, 0, 0, 0, 0, 0, 0, 0, 88, 0, 0, 0, 0, 0, 0, 0
]

# 计算 proof 并输出为字节数组
proof = find_proof(sender_bytes, challenge_bytes)
print(f"Calculated proof as byte array: {proof}")

File: mover/ydamw22/task8/src/getProof.ts
import { sha3_256 } from 'js-sha3';
import { SuiClient, getFullnodeUrl } from '@mysten/sui/client';
import { bcs, fromHex, toHex } from '@mysten/bcs';

const crypto = window.crypto;
const suiClient = new SuiClient({ url: getFullnodeUrl('testnet') });
const objectId = "0x19e76ca504c5a5fa5e214a45fca6c058171ba333f6da897b82731094504d5ab9";
const objectData = await suiClient.getObject({
  id: objectId,
  options: { showContent: true },
});
// 获取testnet上实时challenge对象的数据
const challengeObjectFields = objectData.data?.content?.fields;

// 定义UID
const UID = bcs.fixedArray(32, bcs.u8()).transform({
	input: (id: string) => fromHex(id),
	output: (id) => toHex(Uint8Array.from(id)),
});
// 定义challenge结构
const Challenge = bcs.struct('Challenge', {
	id: UID,
  str: bcs.string(),
  difficulity: bcs.u64(),
  ture_num: bcs.u64(),
});
// 用拿到的链上数据进行bcs序列化
const challengeBytes = Challenge.serialize({
	id: challengeObjectFields.id.id,
  str: challengeObjectFields.str,
  difficulity: challengeObjectFields.difficulity,
  ture_num: challengeObjectFields.ture_num,
}).toBytes();

// 地址转化成bytes
const Address = bcs.bytes(32).transform({
  // To change the input type, you need to provide a type definition for the input
  input: (val: string) => fromHex(val),
  output: (val) => toHex(val),
});
const sender = '0xaf795c762d38591318bf7c471552ce9034be41552eb3607c382776b3b9e68859';
const senderBytes = Address.serialize(sender).toBytes();

// 查找有效proof
export async function findValidProof() {
  // 计数器（5k倍）
  let count = 0;
  for (let i = 0; i < 5000; i++) {
    // 生成8字节数组
    const proofBytes = new Uint8Array(8);
    // 填充随机字节
    crypto.getRandomValues(proofBytes);

    // 拼接完整数据
    const fullProof = full_proof_concatUint8Arrays([
      proofBytes,
      senderBytes,
      challengeBytes
    ]);

    // 计算哈希
    const hash = new Uint8Array(sha3_256.arrayBuffer(fullProof));

    // 验证前3字节和
    const sum = hash.slice(0, 3).reduce((a, b) => a + b, 0);
    if (sum === 0) {
      console.log(`Found valid proof after ${count} attempts`);
      return proofBytes;
    }
    count += 5;
  }
}

// 获取full_proof,拼接Uint8Array
function full_proof_concatUint8Arrays (arrays: Uint8Array[]) {
  let totalLength = arrays.reduce((acc, arr) => acc + arr.length, 0);
  let result = new Uint8Array(totalLength);
  let offset = 0;
  for ( let arr of arrays ) {
    result.set(arr, offset);
    offset += arr.length;
  }
  return result;
}